ui/src/environments/config-loader.service.ts

```diff
@@ -0,0 +1,59 @@
+import { Injectable } from '@angular/core';
+
+@Injectable()
+export class ConfigLoaderService {
+  private environmentConfig: object;
+  private environmentConfigPromise: Promise<any>;
+
+  constructor() { }
+
+  private static loadJSONAsync(rawText:string): Promise<object> {
+    return Promise.resolve(JSON.parse(rawText));
+  }
+
+  getEnvironmentConfig(): Promise<any> {
+    if (!this.environmentConfigPromise) {
+      this.environmentConfigPromise = new Promise<any>((resolve, reject) => {
+        const xmr = new XMLHttpRequest(),
+          method = 'GET',
+          url = '../../../assets/environments/environment.json';
+
+        xmr.open(method, url, true);
+
+        xmr.onload = () => {
+          if (xmr.status === 200) {
+            ConfigLoaderService.loadJSONAsync(xmr.responseText)
+              .catch(SyntaxError => console.log(SyntaxError))
```

Comment:
It's a surprise that when I was trying to take it out, the outer promise cannot catch the SyntaxError raised by the `JSON.parse()`. The current way here is not working either. As it's the latest commit, I have to use `return Promise.resolve(rawText).then(JSON.parse)` and keep this inner `catch` here to catch the SyntaxError properly. 

This observation also makes me think if I have abused using Promise, and should have used the callback styled simple `try..catch(SyntaxError)` for `JSON.parse()`here. What do you think? @bfcrampton

Reply:
Ah okay. I think its just because of the complexity of the promises in this case which is required due to the weird interface on `XMLHttpRequest` (namely setting the `onload` and `onerror` callbacks manually inside of a promise which would be more naturally handled by a `Promise`.

After looking into it more, I think you want to use the [Angular Http Module](https://medium.com/google-developer-experts/angular-2-introduction-to-new-http-module-1278499db2a0), like we do in the `job-manager.service` and `capabilities.service`. I think it would look something like:
```
constructor(private http: Http) {}

getEnvironmentConfig(): Promise<any> {
  return this.http.get(`../../../assets/environments/environment.json`,
    {},
    new RequestOptions({headers: new Headers({'Content-Type': 'application/json'})}))
    .toPromise()
    .then(response => response.json())
    .catch((e) => this.handleError(e));
}
```

The http client also does the parsing into an object for you so you won't even have to parse raw text.

Let me know if something like that works.

Reply:
This approach looks elegant to me, just tried but ran into 2 issues:
1. According to https://angular.io/api/http/Http, the `Http` has been deprecated in Angular. 
2. The deprecation of the `Http` is OK since we can either use the new `HttpClient` to do the something or keep using it like we did in the other services, the tricky problem here is, since the `ConfigLoaderService` has to be instantiated in the main.ts:

If we want to use `Http` since there are no modules stand with main.ts, instead of simply passing `Http` or `HttpModule` in module.ts, we have to manually resolve and create the provider for the Http instance that will be passed into the constructor .e.g `new ConfigLoaderService(injector.get(Http));`, to achieve this, we need something like:
```
const injector = ReflectiveInjector.resolveAndCreate([HTTP_PROVIDERS]);
const configLoader = new ConfigLoaderService(injector.get(Http));
```
However, since `HTTP_PROVIDERS` is also deprecated and *removed* (https://github.com/angular/angular/pull/10864), I have to something like:
```
const HTTP_PROVIDERS = [
  {provide: Http, useFactory:
      (xhrBackend: XHRBackend, requestOptions: RequestOptions): Http =>
        new Http(xhrBackend, requestOptions),
    deps: [XHRBackend, RequestOptions]},
  BrowserXhr,
  {provide: RequestOptions, useClass: BaseRequestOptions},
  {provide: ResponseOptions, useClass: BaseResponseOptions},
  XHRBackend,
  {provide: XSRFStrategy, useFactory: () => new CookieXSRFStrategy()},
];
const injector = ReflectiveInjector.resolveAndCreate(HTTP_PROVIDERS);
const configLoader = new ConfigLoaderService(injector.get(Http));
```
which worked but really clunky.

Another way could be using the new [`HttpClient`](https://angular.io/guide/http), but using it will bring code style inconsistency between the ConfigloaderService and other services(which are still using Http), I just created a ticket regarding this in the board. Given the effort required fo refactoring this, would you mind I treat this as a follow-up PR/work? Or, did i miss anything here? @bfcrampton

Reply:
Ah okay I didn't think about all the dependency injection you'd need since it's in `main.ts`. Definitely fine leaving this as follow up work, and I agree we should move to `HttpClient` over the deprecated `Http` everywhere. 

Does that clunky version work fine and simplify the logic here? I think I'd prefer that if it's easy just because it is more consistent with the rest of the app. But if that also takes work and we're going to migrate over to `HttpClient` later I'm fine with using `XMLHttpRequest` directly.